src/hardware/pcspeaker_impulse.cpp

```diff
@@ -0,0 +1,543 @@
+	static std::vector<float> in_buffer(num_channels);
+	static std::vector<int16_t> out_buffer(num_channels);
+	static auto &out_sample = out_buffer.front();
+	static auto &in_sample  = in_buffer.front();
+
+	while (requested_frames > 0 && waveform_deque.size()) {
+		// Pop the first sample off the waveform
+		in_sample += waveform_deque.front();
+		waveform_deque.pop_front();
+		waveform_deque.push_back(0.0f);
+
+		// Soft-limit the sample, if needed
+		soft_limiter->Process(in_buffer, 1, out_buffer);
+
+		// Pass the sample to the mixer and decrement the requested frames
+		channel->AddSamples_m16(1, &out_sample);
+		--requested_frames;
+
+		// Keep a tally of sequential silence so we can sleep the channel
+		tally_of_silence = out_sample ? 0 : tally_of_silence + 1;
+
+		// Scale down the running volume amplitude. Eventually it will
+		// hit 0 if no other waveforms are generated.
+		in_sample *= high_pass_amount;
+	}
+
+	// Write silence if the waveform deque ran out
+	while (requested_frames) {
+		channel->AddSamples_m16(1, &neutral_amplitude);
+		++tally_of_silence;
+		--requested_frames;
+	}
+
+	// Maybe put the channel to sleep after 10s
+	if (tally_of_silence > 100000) {
+		pit.prev_amplitude = neutral_amplitude;
+		channel->Enable(false);
+		tally_of_silence = 0;
+	}
+}
+
+void PcSpeakerImpulse::InitializeImpulseLUT()
+{
+	assert(impulse_lut.size() == filter_width);
+	for (auto i = 0u; i < filter_width; ++i)
+		impulse_lut[i] = CalcImpulse(i / (static_cast<double>(sample_rate) * oversampling_factor));
+}
+
+void PcSpeakerImpulse::SetFilterState(const FilterState filter_state)
+{
+	assert(channel);
+
+	// Setup filters
+	if (filter_state == FilterState::On) {
+		// The filters are meant to emulate the bandwidth limited
+		// sound of the small PC speaker. This more accurately
+		// reflects people's actual experience of the PC speaker
+		// sound than the raw unfiltered output, and it's a lot
+		// more pleasant to listen to, especially in headphones.
+		constexpr auto hp_order       = 3;
+		constexpr auto hp_cutoff_freq = 120;
+		channel->ConfigureHighPassFilter(hp_order, hp_cutoff_freq);
+		channel->SetHighPassFilter(FilterState::On);
+
+		constexpr auto lp_order       = 3;
+		constexpr auto lp_cutoff_freq = 4300;
+		channel->ConfigureLowPassFilter(lp_order, lp_cutoff_freq);
+		channel->SetLowPassFilter(FilterState::On);
+	} else {
+		channel->SetHighPassFilter(FilterState::Off);
+		channel->SetLowPassFilter(FilterState::Off);
+	}
+}
+
+PcSpeakerImpulse::PcSpeakerImpulse()
+{
+	static_assert(sample_rate >= 8000, "Sample rate must be at least 8 kHz");
+	static_assert(sample_rate % 1000 == 0, "PC Speaker sample must be a multiple of 1000");
+
+	InitializeImpulseLUT();
+
+	// Size the waveform queue
+	constexpr auto waveform_size = filter_quality + sample_rate_per_ms;
+	waveform_deque.resize(waveform_size, 0.0f);
+
+	// Setup the soft limiter
+	soft_limiter = std::make_unique<SoftLimiter>(device_name);
+	assert(soft_limiter);
```

Comment:
No longer needed/useful with the float mixer path, as discussed. Could you remove it?

Reply:
Gone!